src/puppetlabs/puppetdb/mq_listener.clj

```diff
@@ -339,25 +344,143 @@
+                session (create-session connection)]
+      (let [q (create-queue session endpoint)]
+        (with-open [producer (.createProducer session q)]
+
+          (.send producer (mq/to-jms-message session message-to-convert message-properties)))))))
+
+(defn create-command-consumer
+  "Create and return a command handler. This function does the work of
+  consuming/storing a command. Handled commands are acknowledged here"
+  [{:keys [discard-dir] :as mq-context} command-handler]
+  (let [handle-message (wrap-message-handler-middleware (send-delayed-message mq-context)
+                                                        discard-dir
+                                                        command-handler)]
+    (fn [^Message message]
+      ;; When the queue is shutting down, it sends nil message
+      (when message
+        (try
+          (handle-message (mq/convert-jms-message message))
+          (.acknowledge message)
+          (catch Exception ex
+            (.rollback message)
+            (log/error ex "Unable to process message. Message not acknowledged and will be retried")))))))
+
+(defn create-mq-receiver
+  "Infinite loop, intended to be run in an isolated thread that
+  recieves a message from ActiveMQ, passes it to `mq-message-handler`
+  and looks for the next message"
+  [{:keys [^ConnectionFactory conn-pool endpoint]} mq-message-handler mq-connected-promise]
+  (with-open [connection (create-connection conn-pool)
+              session (create-session connection)
+              consumer (.createConsumer session (create-queue session endpoint))]
+    (try
+      ;; delivering the promise here we have ensured that we can
+      ;; connect to the MQ and that the queue has been created (for
+      ;; the first startup of PuppetDB)
+      (deliver mq-connected-promise true)
+      (loop []
+        (mq-message-handler (.receive consumer))
+        (recur))
+      (catch javax.jms.IllegalStateException e
+        (log/info "Received IllegalStateException, shutting down"))
+      (catch Exception e
+        (log/error e)))))
+
+(def logging-exception-handler
+  "Exception handler that ensures any uncaught exception that occurs
+  on this thread is logged"
+  (reify Thread$UncaughtExceptionHandler
+    (uncaughtException [_ thread throwable]
+      (log/error throwable "Error processing command"))))
+
+(defn command-thread-factory
+  "Creates a command thread factory, wrapping the
+  `pool-thread-factory`. Using this thread factory ensures that we
+  have distinguishable names for threads in this pool and that all
+  uncaught exceptions are logged as errors. Without this uncaught
+  exceptions get the default which is standard error"
+  [pool-thread-factory]
+  (-> (BasicThreadFactory$Builder.)
+      (.wrappedFactory pool-thread-factory)
+      (.namingPattern "cmd-proc-thread-%d")
+      (.uncaughtExceptionHandler logging-exception-handler)
+      (.daemon true)
```

Comment:
Perhaps we already know this is appropriate/safe, but generally speaking, I'm a little wary since (I think) daemon threads can be abruptly killed at shutdown, i.e. no finally-blocks/stack-unwinding, etc.  So if we're not sure, then I'd suggest we just make sure our threads can be (safely) stopped - whether via explicit time-to-die? checking of some kind, or possibly carefully arranged interruption, and then wait for them to quit.